src/nupic/encoders/Scalar.cpp

```diff
@@ -0,0 +1,248 @@
+      break;
+    case NTA_BasicType_UInt64:
+      return ((UInt64*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Real32:
+      return ((Real32*) input.getBuffer())[0];
+      break;
+    case NTA_BasicType_Real64:
+      return ((Real64*) input.getBuffer())[0];
+      break;
+    default:
+      NTA_THROW << "extractScalar: Unsupported type " << input.getType();
+    }
+  }
+
+  ScalarEncoder::ScalarEncoder(
+    int w, double minValue, double maxValue, int n, double radius,
+    double resolution, bool clipInput)
+    :w_(w),
+     minValue_(minValue),
+     maxValue_(maxValue),
+     clipInput_(clipInput)
+  {
+    if ((n != 0 && (radius != 0 || resolution != 0)) ||
+        (radius != 0 && (n != 0 || resolution != 0)) ||
+        (resolution != 0 && (n != 0 || radius != 0)))
+    {
+      NTA_THROW <<
+        "Only one of n/radius/resolution can be specified for a ScalarEncoder.";
+    }
+
+    const double extentWidth = maxValue - minValue;
+    if (extentWidth <= 0)
+    {
+      NTA_THROW << "minValue must be < maxValue. minValue=" << minValue <<
+        " maxValue=" << maxValue;
+    }
+
+    if (n != 0)
+    {
+      n_ = n;
+
+      if (w_ < 1 || w_ >= n_)
+      {
+        NTA_THROW << "w must be within the range [1, n). w=" << w_ << " n=" << n_;
+      }
+
+      // Distribute nBuckets points along the domain [minValue, maxValue],
+      // including the endpoints. The resolution is the width of each band
+      // between the points.
+      const int nBuckets = n - (w - 1);
+      const int nBands = nBuckets - 1;
+      bucketWidth_ = extentWidth / nBands;
+    }
+    else
+    {
+      bucketWidth_ = resolution || radius / w;
+      if (bucketWidth_ == 0)
+      {
+        NTA_THROW << "One of n/radius/resolution must be nonzero.";
+      }
+
+      const int neededBands = ceil(extentWidth / bucketWidth_);
+      const int neededBuckets =  neededBands + 1;
+      n_ = neededBuckets + (w - 1);
+    }
+  }
+
+  ScalarEncoder::~ScalarEncoder()
+  {
+  }
+
+  void ScalarEncoder::encodeIntoArray(
+    const ArrayBase & input, UInt output[], bool learn)
+  {
+    Real64 scalar = extractScalar(input);
+
+    if (scalar < minValue_)
+    {
+      if (clipInput_)
+      {
+        scalar = minValue_;
+      }
+      else
+      {
+        NTA_THROW << "input (" << scalar << ") less than range [" << minValue_ <<
+          ", " << maxValue_ << "]";
+      }
+    } else if (scalar > maxValue_) {
+      if (clipInput_)
+      {
+        scalar = maxValue_;
+      }
+      else
+      {
+        NTA_THROW << "input (" << scalar << ") greater than range [" << minValue_ <<
+          ", " << maxValue_ << "]";
+      }
+    }
+
+    const int iBucket = round((scalar - minValue_) / bucketWidth_);
+
+    const int firstBit = iBucket;
+
+    memset(output, 0, n_*sizeof(output[0]));
+    for (int i = 0; i < w_; i++)
+    {
+      output[firstBit + i] = 1;
+    }
+  }
+
+  PeriodicScalarEncoder::PeriodicScalarEncoder(
```

Comment:
Basically, I think that logic should live at the bridge level. If someone wants to bridge these classes to other languages via SWIG, the SWIG layer can expose a common ScalarEncoder constructor which chooses which type to instantiate at runtime.

Same for defining networks via dictionaries (a.k.a. model params). Same for serialization.

We should give the implementations some freedom to refactor. The "let's have a common interface" goal should be achieved at a different layer in the code.

Note that I'm working on getting hotgym running on the C++ encoders, and having low-friction switching between Python and C++ implementations is a priority. If I'm wrong, it'll be easy to re-entangle the two classes.